firewalldb/sql_migration_test.go

```diff
@@ -0,0 +1,527 @@
+				)
+				require.NoError(t, err)
+
+				featureIDs[featureName] = featureID
+			}
+
+			return featureID
+		}
+
+		// First we extract all migrated kv records from the SQLDB,
+		// in order to be able to compare them to the original kv
+		// records, to ensure that the migration was successful.
+		sqlKvRecords, err := sqlStore.ListAllKVStoresRecords(ctx)
+		require.NoError(t, err)
+		require.Equal(t, len(kvRecords), len(sqlKvRecords))
+
+		for _, kvRecord := range kvRecords {
+			ruleID := getRuleID(kvRecord.RuleName)
+
+			if kvRecord.Global {
+				sqlVal, err := sqlStore.GetGlobalKVStoreRecord(
+					ctx,
+					sqlc.GetGlobalKVStoreRecordParams{
+						Key:    kvRecord.EntryKey,
+						Perm:   kvRecord.Perm,
+						RuleID: ruleID,
+					},
+				)
+				require.NoError(t, err)
+				require.Equal(t, kvRecord.Value, sqlVal)
+			} else if kvRecord.FeatureName.IsNone() {
+				groupID := getGroupID(kvRecord.GroupID[:])
+
+				sqlVal, err := sqlStore.GetSessionKVStoreRecord(
+					ctx,
+					sqlc.GetSessionKVStoreRecordParams{
+						Key:    kvRecord.EntryKey,
+						Perm:   kvRecord.Perm,
+						RuleID: ruleID,
+						SessionID: sql.NullInt64{
+							Int64: groupID,
+							Valid: true,
+						},
+					},
+				)
+				require.NoError(t, err)
+				require.Equal(t, kvRecord.Value, sqlVal)
+			} else {
+				groupID := getGroupID(kvRecord.GroupID[:])
+				featureID := getFeatureID(
+					kvRecord.FeatureName.UnwrapOrFail(t),
+				)
+
+				sqlVal, err := sqlStore.GetFeatureKVStoreRecord(
+					ctx,
+					sqlc.GetFeatureKVStoreRecordParams{
+						Key:    kvRecord.EntryKey,
+						Perm:   kvRecord.Perm,
+						RuleID: ruleID,
+						SessionID: sql.NullInt64{
+							Int64: groupID,
+							Valid: true,
+						},
+						FeatureID: sql.NullInt64{
+							Int64: featureID,
+							Valid: true,
+						},
+					},
+				)
+				require.NoError(t, err)
+				require.Equal(t, kvRecord.Value, sqlVal)
+			}
+		}
+	}
+
+	// The tests slice contains all the tests that we will run for the
+	// migration of the firewalldb from a BoltDB to a SQLDB.
+	// Note that the tests currently only test the migration of the KV
+	// stores, but will be extended in the future to also test the migration
+	// of the privacy mapper and the actions store.
+	tests := []struct {
+		name       string
+		populateDB func(t *testing.T, ctx context.Context,
+			boltDB *BoltDB,
+			sessionStore session.Store) []kvStoreRecord
+	}{
+		{
+			name: "empty",
+			populateDB: func(t *testing.T, ctx context.Context,
+				boltDB *BoltDB,
+				sessionStore session.Store) []kvStoreRecord {
+
+				// Don't populate the DB.
+				return make([]kvStoreRecord, 0)
+			},
+		},
+		{
+			name:       "global records",
+			populateDB: globalRecords,
+		},
+		{
+			name:       "session specific records",
+			populateDB: sessionSpecificRecords,
+		},
+		{
+			name:       "feature specific records",
+			populateDB: featureSpecificRecords,
+		},
+		{
+			name:       "random records",
+			populateDB: randomKVRecords,
```

Comment:
i think it's worth having a case that does some explicit entries that write to all the store types. currently we only do that for the random one.

Reply:
Good idea! Added an extra test with the latest push.

Reply:
Ah yeah, good point! I updated this test quite a bit (now renamed to `allEntryCombinations`) to hopefully cover all different types of entries in a deterministic way. I think it should now cover all of those cases.